source/grid/tria.cc

```diff
@@ -14479,7 +14481,9 @@ Triangulation<dim, spacedim>::prepare_coarsening_and_refinement()
                     // only do something if the face is not at the
                     // boundary and if the face will be refined with
                     // the RefineCase currently flagged for
-                    if (cell->neighbor(i).state() == IteratorState::valid &&
+                    const bool has_periodic_neighbor =
+                      cell->has_periodic_neighbor(i);
+                    if ((!cell->at_boundary(i) || has_periodic_neighbor) &&
```

Comment:
this condition "not at boundary, or has periodic neighbor" took me quite a while to understand. I think it would be easier to read if you did
```
  const bool has_neighbor_or_periodic_neighbor = ...;
  if (has_neighbor_or_periodic_neighbor && ...
```